lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -1828,7 +2153,31 @@ def main():
             docker_api_version='1.25',
             detect_usage=_detect_healthcheck_start_period,
             usage_msg='set healthcheck.start_period'
-        )
+        ),
+        update_config_max_failure_ratio=dict(
+            docker_py_version='2.1.0',
+            docker_api_version='1.25',
+            detect_usage=lambda c: c.module.params['update_config']['max_failure_ratio'] is not None,
+            usage_msg='set update_config.max_failure_ratio'
+        ),
+        update_config_monitor=dict(
+            docker_py_version='2.1.0',
+            docker_api_version='1.25',
+            detect_usage=lambda c: c.module.params['update_config']['monitor'] is not None,
+            usage_msg='set update_config.monitor'
+        ),
+        update_config_order=dict(
+            docker_py_version='2.7.0',
+            docker_api_version='1.29',
+            detect_usage=lambda c: c.module.params['update_config']['order'] is not None,
+            usage_msg='set update_config.order'
+        ),
+        placement_config_preferences=dict(
+            docker_py_version='2.4.0',
+            docker_api_version='1.27',
+            detect_usage=lambda c: c.module.params['placement']['preferences'] is not None,
+            usage_msg='set placement.preferences'
+        ),
```

Comment:
I think below (after the `AnsibleDockerClient` initialization) it would be good to have checks for the old options: in case they are used, issue a warning that users should switch to the new option; you can see in https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_swarm.py#L401-L403 how such a warning can be emitted. That will make the deprecation more visible and make sure users don't forget to change this before their playbooks suddenly stop working with Ansible 2.12 :)

Reply:
Hmm maybe I'm missing something but isn't that what passing `removed_in_version` in the argument spec is already doing?

Reply:
Yes, you are right. I totally forgot about that one... (The only downside is that it doesn't tell which option to use instead :) )